lib/min_heap.rb

```diff
@@ -44,28 +48,46 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    return if index == 0
+    parent_index = (index - 1) / 2
+    if @store[index].key < @store[parent_index].key
+      swap(index, parent_index)
+      heap_up(parent_index)
+    end
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    left_child = (index * 2) + 1
+    right_child = (index * 2) + 2
+    
+    return if @store[left_child].nil? || @store[right_child].nil? 
```

Comment:
What about if the left child is **not** `nil` and the right child is?  

Could you end up not making a swap you needed to?  This is a rare but possible scenario.